Doberman/Pipeline.py

```diff
@@ -113,45 +121,73 @@ def build(self, config):
-                self.logger.debug(f'Created {nodes_built} nodes this iter, {len(self.graph)}/{len(pipeline_config)} total')
+                self.logger.debug(f'Created {nodes_built} nodes this iter, {len(graph)}/{len(pipeline_config)} total')
         for kwargs in pipeline_config:
             for u in kwargs.get('upstream', []):
-                self.graph[u].downstream_nodes.append(self.graph[kwargs['name']])
+                graph[u].downstream_nodes.append(graph[kwargs['name']])
 
         self.startup_cycles = num_buffer_nodes + longest_buffer # I think?
         self.logger.debug(f'I estimate we will need {self.startup_cycles} cycles to start')
 
+        self.calculate_jointedness(self, graph)
+
+    def calculate_jointedness(self, graph):
+        """
+        Takes in the graph as created above and figures out how many
+        disjoint sections it has. These sections get separated out into subpipelines
+        """
+        while len(graph):
+            self.logger.debug(f'{len(graph)} nodes to check')
+            nodes_to_check = set([list(graph.keys())[0]])
+            nodes_checked = set()
+            pl = {}
+            while len(nodes_to_check) != 0:
+                node = nodes_to_check.pop()
+                self.logger.debug(f'Checking {node}')
+                for u in graph[node].upstream_nodes:
+                    if u.name not in nodes_checked:
+                        nodes_to_check.add(u.name)
+                for d in graph[node].downstream_nodes:
+                    if d.name not in nodes_checked:
+                        nodes_to_check.add(d.name)
+                pl[node] = graph.pop(node)
+                nodes_checked.add(node)
+            self.logger.debug(f'Found set: {set(pl.keys()}')
+            self.subpipelines.append(pl)
+
     def reconfigure(self, doc):
-        for node in self.graph.values():
-            if isinstance(node, Doberman.AlarmNode):
-                rd = self.db.get_sensor_setting(name=node.input_var)
-                if node.name not in doc:
-                    doc[node.name] = {}
-                doc[node.name].update(alarm_thresholds=rd['alarm_thresholds'], readout_interval=rd['readout_interval'])
-            if isinstance(node, Doberman.SimpleAlarmNode):
-                if node.name not in doc:
-                    doc[node.name] = {}
-                doc[node.name].update(length=rd['alarm_recurrence'])
-            if node.name in doc:
-                node.load_config(doc[node.name])
+        for pl in self.subpipelines:
+            for node in pl.values():
+                if isinstance(node, Doberman.AlarmNode):
+                    rd = self.db.get_sensor_setting(name=node.input_var)
+                    if node.name not in doc:
+                        doc[node.name] = {}
+                    doc[node.name].update(alarm_thresholds=rd['alarm_thresholds'], readout_interval=rd['readout_interval'])
+                if isinstance(node, Doberman.SimpleAlarmNode):
```

Comment:
`rd = self.db.get_sensor_setting(name=node.input_var)`

Reply:
I don't think it's necessary. `SimpleAlarmNode` subclasses from `AlarmNode`, so there isn't any way for the L179 block to run if the L174 block hasn't. I adjusted the indentation to make this a bit more clear.

Reply:
oh true.